File: message.php
<?php
session_start();

// Database Connection
$host = "localhost";
$username = "root";
$password = "";
$database = "guidenet";
$conn = new mysqli($host, $username, $password, $database);

if ($conn->connect_error) {
    die("Connection failed: " . $conn->connect_error);
}

$id = $_SESSION['id'];

// Fetch user details
$query = "SELECT * FROM stud_registration WHERE ID = ?";
$stmt = $conn->prepare($query);
$stmt->bind_param("s", $id);
$stmt->execute();
$result = $stmt->get_result();

if ($result->num_rows > 0) {
    $user = $result->fetch_assoc();
} else {
    echo "User details not found!";
    exit();
}

$get_name_query = "SELECT NAME FROM stud_registration WHERE ID = '$id'";
$result = mysqli_query($conn, $get_name_query);
$row = mysqli_fetch_assoc($result);
$senderName = $row['NAME']; // Fetch sender's name

?>

<!DOCTYPE html>
<html lang="en">
<head>
    <meta charset="UTF-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>Chat Portal</title>
    <link rel="stylesheet" href="style_message.css">
    <script src="script.js" defer></script>
</head>
<body>
    <div class="container">
        <div class="user-list">
            <h3>Users</h3>
            <ul id="user-list"></ul>
        </div>
        <div class="chat-box">
            <h3 id="chat-header">Chat Area</h3>
    
            <script>
            // Store sender name in localStorage
            localStorage.setItem("id", "<?php echo $id; ?>");
            localStorage.setItem("name", "<?php echo $senderName; ?>");

            let loggedInUserId = sessionStorage.getItem("userId");
            function openChat(userId, userName) {
                //if (userId == loggedInUserId) return;

                // Update chat header dynamically
                document.getElementById("chat-header").innerText = "Chat with " + userName;
            
                // Fetch and display previous messages
                fetch(`http://127.0.0.1:5000/get_messages?receiver_id=${userId}`)
                .then(response => response.json())
                .then(messages => {
                    let chatBox = document.getElementById("chat-box");
                    chatBox.innerHTML = ""; // Clear old messages
                    messages.forEach(msg => {
                        let p = document.createElement("p");
                        p.innerText = msg.sender + ": " + msg.message;
                        chatBox.appendChild(p);
                    });
                })
                .catch(error => console.error("Error fetching messages:", error));
            }
            </script>
        
            <div id="chat-area" class="chat-messages"></div>
            <div class="input-area">
                <input type="text" id="message-input" placeholder="Type Message">
                <button onclick="sendMessage()">Send</button>
            </div>
        </div>
    </div>
</body>
</html>

File: message.py
from flask import Flask, request, jsonify, render_template
import mysql.connector  
from flask_cors import CORS  # Import CORS

app = Flask(__name__)
CORS(app)  # Initialize CORS

# Database Connection
def get_db_connection():
    return mysql.connector.connect(
        host="localhost",
        user="root",
        password="",
        database="guidenet"
    )
db = get_db_connection()
cursor = db.cursor()

# API to Fetch Users
@app.route('/get_users', methods=['GET'])
def get_users():
    logged_in_user_id = request.args.get("id")
    db = get_db_connection()
    cursor = db.cursor()
    try:
        cursor.execute("SELECT id, name FROM stud_registration where id != %s",(logged_in_user_id,) )  # Fetch user ID and Name
        users = [{"id": row[0], "name": row[1]} for row in cursor.fetchall()]
        print("Fetched users:", users)  # Debugging output in terminal
        cursor.close()
        db.close()
        return jsonify(users) 
        
    except Exception as e:
        print("Database error:", str(e))
        return jsonify({"error": str(e)}), 500


@app.route('/get_messages', methods=['GET'])
def get_messages():
    sender_id = request.args.get('sender_id')
    receiver_id = request.args.get('receiver_id')
    db = get_db_connection()
    cursor = db.cursor()
    query = "SELECT sender_id, message FROM messages WHERE (sender_id = %s AND receiver_id = %s) OR (sender_id = %s AND receiver_id = %s) ORDER BY timestamp"
    cursor.execute(query, (sender_id, receiver_id, receiver_id, sender_id))
    
    messages = [{"sender": row[0], "message": row[1]} for row in cursor.fetchall()]
    cursor.close()
    db.close()
    return jsonify(messages)


@app.route('/send_message', methods=['POST'])
def send_message():
    try:
        data = request.get_json()
        print("Received data:", data)

        senderName = request.json.get('senderName')
        if not senderName:
            return jsonify({"error": "Sender name not found"}), 400

        sender_id = data.get('sender_id')
        receiver_id = data['receiver_id']
        message = data['message']

        db = get_db_connection()
        cursor = db.cursor()

        # 🔹 Fetch sender's name dynamically
        cursor.execute("SELECT NAME FROM stud_registration WHERE ID = %s", (sender_id,))
        senderName_result = cursor.fetchone()

        if senderName_result:
            senderName = senderName_result[0]  # Extract name
        else:
            return jsonify({"success": False, "error": "Sender name not found"}), 400

        query = "INSERT INTO messages (sender_name, sender_id, receiver_id, message) VALUES (%s, %s, %s, %s)"
        cursor.execute(query, (senderName, sender_id, receiver_id, message))
        db.commit()
        
        cursor.close()
        db.close()

        return jsonify({"success": True, "message": "Message sent successfully!"})

    except Exception as e:
        return jsonify({"success": False, "error": str(e)}), 500

if __name__ == '__main__':
    app.run(debug=True)
